CI: Reduce HTTP requests when possible

jserv opened the pull request:
Due to frequent CI failures during checksum verification, this commit bypasses integrity checks by assuming all downloaded files are complete, removing the requirement to download and verify checksums. 
 <div id='description'>
<h3>Summary by Bito</h3>
This PR enhances the CI build process by optimizing HTTP requests and file download handling through the introduction of a fetch-releases-tag function. The changes include improved file existence checks to prevent redundant downloads, streamlined build processes with better checksum verification, and enhanced cleanup procedures in the Makefile.
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
</div>

jserv commented:
@vacantron, I tried to bypass checksum validation, but it did not work as intended. Can you help me fix this?

otteryc commented on mk/artifact.mk:
```diff
@@ -60,25 +60,35 @@ endif
 	$(Q)$(eval RES := 0)
+else
+	# Assume every downloaded file is verified.
+	$(Q)$(eval RES := 1)
```
Setting `RES` to 1 here actually indicates an abnormal event, i.e., the SHA-1 checksum verification has failed or the file does not exist.
Originally, the variable `RES` is passed into the macro `verify` located in `mk/external.mk`. However, the `verify` process is bypassed when `SKIP_CHECKSUM ` is set. Consequently, the `RES` variable remains unchanged, resulting in the perpetual re-download of the prebuilt file whenever the target `artifact` is triggered.

otteryc commented on mk/artifact.mk:
```diff
@@ -107,6 +118,9 @@ else
 	fi
 endif
+ifneq ($(SKIP_CHECKSUM),)
+	$(Q)$(eval RES := 0)
```
My apologies for any confusion.
Setting `RES` to `1` wrongly makes that the error-checking statements at lines 107 and 114 are always true. This implies that it is incorrect to set `RES` at a previous position.

Furthermore, omitting the entire `fetch-checksum` process may lead to a predicament, as the filenames, which  should be downloaded, are retrieved from the checksum file. Additionally, we cannot verify the file’s existence by the tarball, as it is directly piped into the `tar` utility. Consequently, it is not possible to cover the scenario that the files do not exist in current implementation.

One potential solution is to store the prebuilt tarball instead of piping it, thereby ensuring its verification.
For reference, please refer to the following commit: https://github.com/otteryc/rv32emu/commit/5cf1e728e46a7070325e6eca32e79e81c6b4a4f7. This test run verifies the existence of the tarball, although the CI test (clang) still fails due to exceeding the rate limit while downloading the prebuilt tarball.

vacantron committed:
CI: Reduce HTTP requests

Check prebuilt tarballs are existing or not first. If so, skip the
downloading of releases tags and tarballs.

In the CI, we first call "make artifact" to prepare the tarballs, so the
GitHub API for fetching releases tags would not be called and be skipped
when "make" is called next time.

After this commit, wget in the CI would not complain about error code
403 (rate limit exceeded) anymore. However, the build on Aarch64 still
failed because of other reansons, e.g. software bugs of GCC or QEMU.

Close #536

jserv commented:
I defer to @vacantron to determine if we should proceed with the proposed changes.

jserv commented:
Append `Close #536` at the end of git commit messages.

jserv commented:
Mention Arm64 regressions in git commit messages to enable earlier pull request merging despite ongoing CI pipeline issues.

jserv commented on mk/artifact.mk:
```diff
@@ -37,14 +37,27 @@ SCIMARK2_SHA1 := de278c5b8cef84ab6dda41855052c7bfef919e36
-  PREBUILT_BLOB_URL = https://github.com/sysprog21/rv32emu-prebuilt/releases/download/$(LATEST_RELEASE)
+    ifeq ($(call has, SYSTEM), 1)
+        ifeq ($(wildcard $(BIN_DIR)/rv32emu-linux-image-prebuilt.tar.gz),)
+            LATEST_RELEASE := $(shell wget -q https://api.github.com/repos/sysprog21/rv32emu-prebuilt/releases -O- | grep '"tag_name"' | grep "Linux-Image" | head -n 1 | sed -E 's/.*"tag_name": "([^"]+)".*/\1/')
```
Rewrite as following:
```shell
wget -qO- https://api.github.com/repos/sysprog21/rv32emu-prebuilt/releases \
  | grep -m1 '"tag_name".*Linux-Image' \
  | sed -E 's/.*"tag_name": "([^"]+)".*/\1/'
```
and consider to use `define` for code reusability.

jserv commented on mk/artifact.mk:
```diff
@@ -149,14 +163,26 @@ fetch-checksum:
-	$(Q)$(call notice, [OK])
+    ifeq ($(wildcard $(BIN_DIR)/rv32emu-prebuilt.tar.gz),)
+		wget -q -O $(BIN_DIR)/sha1sum-linux-x86-softfp $(PREBUILT_BLOB_URL)/sha1sum-linux-x86-softfp
+		wget -q -O $(BIN_DIR)/sha1sum-riscv32 $(PREBUILT_BLOB_URL)/sha1sum-riscv32
```
Later, we may consider to introduce [function calls](https://www.gnu.org/software/make/manual/html_node/Call-Function.html) to wrap `wget` in another pull request.

vacantron replied:
Sure. Moreover, consolidating the file name of sha1sum and prebuilts, such as `rv32emu-prebuilt-*.sha` would make it easier.

ChinYikMing commented on .github/workflows/main.yml:
```diff
@@ -58,10 +58,20 @@ jobs:
+      run: |
+            make artifact
+            make ENABLE_SYSTEM=1 artifact
+            make ENABLE_ARCH_TEST=1 artifact
```
Since `LATEST_RELEASE` is guarded to the tarball, I am not sure if fetching all artifacts in a single job is a good choice.

Would not it make more sense to fetch the required artifact in a relevant job instead? If it is not, the line 242 should be refined such that removing the redundant artifact fetching.

vacantron replied:
This is just moving up the artifacts fetching. We do all userspace test, system simulation test, and architecture test below.

The redundant `make artifact` could not cause any problem. Beccause once the tarball is downloaded, we check that tarball to decide triggering downloading or not.

vacantron commented:
The new patch to solve the Arm64 CI failure will be opened after this pull request merged.

https://github.com/sysprog21/rv32emu/actions/runs/13257429033/job/37006598258?pr=560

jserv commented:
Thank @vacantron for contributing!
